File: swim-system-java/swim-core-java/swim.recon/src/main/java/swim/recon/ReconWriter.java
// Copyright 2015-2019 SWIM.AI inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package swim.recon;

import java.math.BigInteger;
import java.nio.ByteBuffer;
import java.util.Iterator;
import swim.codec.Base10;
import swim.codec.Base16;
import swim.codec.Output;
import swim.codec.Unicode;
import swim.codec.Writer;

/**
 * Factory for constructing Recon writers.
 */
public abstract class ReconWriter<I, V> {
  public abstract boolean isField(I item);

  public abstract boolean isAttr(I item);

  public abstract boolean isSlot(I item);

  public abstract boolean isValue(I item);

  public abstract boolean isRecord(I item);

  public abstract boolean isText(I item);

  public abstract boolean isNum(I item);

  public abstract boolean isBool(I item);

  public abstract boolean isExpression(I item);

  public abstract boolean isExtant(I item);

  public abstract Iterator<I> items(I item);

  public abstract I item(V value);

  public abstract V key(I item);

  public abstract V value(I item);

  public abstract String string(I item);

  public abstract int precedence(I item);

  public abstract int sizeOfItem(I item);

  public abstract Writer<?, ?> writeItem(I item, Output<?> output);

  public abstract int sizeOfValue(V value);

  public abstract Writer<?, ?> writeValue(V value, Output<?> output);

  public abstract int sizeOfBlockValue(V value);

  public abstract Writer<?, ?> writeBlockValue(V value, Output<?> output);

  public int sizeOfAttr(V key, V value) {
    return AttrWriter.sizeOf(this, key, value);
  }

  public Writer<?, ?> writeAttr(V key, V value, Output<?> output) {
    return AttrWriter.write(output, this, key, value);
  }

  public int sizeOfSlot(V key, V value) {
    return SlotWriter.sizeOf(this, key, value);
  }

  public Writer<?, ?> writeSlot(V key, V value, Output<?> output) {
    return SlotWriter.write(output, this, key, value);
  }

  public abstract int sizeOfBlockItem(I item);

  public abstract Writer<?, ?> writeBlockItem(I item, Output<?> output);

  public int sizeOfBlock(Iterator<I> items, boolean inBlock, boolean inMarkup) {
    return BlockWriter.sizeOf(this, items, inBlock, inMarkup);
  }

  public Writer<?, ?> writeBlock(Iterator<I> items, Output<?> output, boolean inBlock, boolean inMarkup) {
    return BlockWriter.write(output, this, items, inBlock, inMarkup);
  }

  public int sizeOfBlock(I item) {
    final Iterator<I> items = items(item);
    if (items.hasNext()) {
      return BlockWriter.sizeOf(this, items, isBlockSafe(items(item)), false);
    } else {
      return 2; // "{}"
    }
  }

  public Writer<?, ?> writeBlock(I item, Output<?> output) {
    final Iterator<I> items = items(item);
    if (items.hasNext()) {
      return BlockWriter.write(output, this, items, isBlockSafe(items(item)), false);
    } else {
      return Unicode.writeString("{}", output);
    }
  }

  public int sizeOfRecord(I item) {
    final Iterator<I> items = items(item);
    if (items.hasNext()) {
      return BlockWriter.sizeOf(this, items, false, false);
    } else {
      return 2; // "{}"
    }
  }

  public Writer<?, ?> writeRecord(I item, Output<?> output) {
    final Iterator<I> items = items(item);
    if (items.hasNext()) {
      return BlockWriter.write(output, this, items, false, false);
    } else {
      return Unicode.writeString("{}", output);
    }
  }

  public int sizeOfPrimary(V value) {
    if (isRecord(item(value))) {
      final Iterator<I> items = items(item(value));
      if (items.hasNext()) {
        return PrimaryWriter.sizeOf(this, items);
      }
    } else if (!isExtant(item(value))) {
      return sizeOfValue(value);
    }
    return 2; // "()"
  }

  public Writer<?, ?> writePrimary(V value, Output<?> output) {
    if (isRecord(item(value))) {
      final Iterator<I> items = items(item(value));
      if (items.hasNext()) {
        return PrimaryWriter.write(output, this, items);
      }
    } else if (!isExtant(item(value))) {
      return writeValue(value, output);
    }
    return Unicode.writeString("()", output);
  }

  public boolean isBlockSafe(Iterator<I> items) {
    while (items.hasNext()) {
      if (isAttr(items.next())) {
        return false;
      }
    }
    return true;
  }

  public boolean isMarkupSafe(Iterator<I> items) {
    if (!items.hasNext() || !isAttr(items.next())) {
      return false;
    }
    while (items.hasNext()) {
      if (isAttr(items.next())) {
        return false;
      }
    }
    return true;
  }

  public int sizeOfMarkupText(I item) {
    return sizeOfMarkupText(string(item));
  }

  public Writer<?, ?> writeMarkupText(I item, Output<?> output) {
    return writeMarkupText(string(item), output);
  }

  public int sizeOfMarkupText(String text) {
    return MarkupTextWriter.sizeOf(text);
  }

  public Writer<?, ?> writeMarkupText(String text, Output<?> output) {
    return MarkupTextWriter.write(output, text);
  }

  public int sizeOfData(int length) {
    return DataWriter.sizeOf(length);
  }

  public Writer<?, ?> writeData(ByteBuffer value, Output<?> output) {
    if (value != null) {
      return DataWriter.write(output, value);
    } else {
      return Unicode.writeString("%", output);
    }
  }

  public boolean isIdent(I item) {
    return isIdent(string(item));
  }

  public boolean isIdent(String value) {
    final int n = value.length();
    if (n == 0 || !Recon.isIdentStartChar(value.codePointAt(0))) {
      return false;
    }
    for (int i = value.offsetByCodePoints(0, 1); i < n; i = value.offsetByCodePoints(i, 1)) {
      if (!Recon.isIdentChar(value.codePointAt(i))) {
        return false;
      }
    }
    return true;
  }

  public int sizeOfText(String value) {
    if (isIdent(value)) {
      return IdentWriter.sizeOf(value);
    } else {
      return StringWriter.sizeOf(value);
    }
  }

  public Writer<?, ?> writeText(String value, Output<?> output) {
    if (isIdent(value)) {
      return IdentWriter.write(output, value);
    } else {
      return StringWriter.write(output, value);
    }
  }

  public int sizeOfNum(int value) {
    int size = Base10.countDigits(value);
    if (value < 0) {
      size += 1;
    }
    return size;
  }

  public Writer<?, ?> writeNum(int value, Output<?> output) {
    return Base10.writeInt(value, output);
  }

  public int sizeOfNum(long value) {
    int size = Base10.countDigits(value);
    if (value < 0L) {
      size += 1;
    }
    return size;
  }

  public Writer<?, ?> writeNum(long value, Output<?> output) {
    return Base10.writeLong(value, output);
  }

  public int sizeOfNum(float value) {
    return Float.toString(value).length();
  }

  public Writer<?, ?> writeNum(float value, Output<?> output) {
    return Base10.writeFloat(value, output);
  }

  public int sizeOfNum(double value) {
    return Double.toString(value).length();
  }

  public Writer<?, ?> writeNum(double value, Output<?> output) {
    return Base10.writeDouble(value, output);
  }

  public int sizeOfNum(BigInteger value) {
    return value.toString().length();
  }

  public Writer<?, ?> writeNum(BigInteger value, Output<?> output) {
    return Unicode.writeString(value, output);
  }

  public int sizeOfUint32(int value) {
    return 10;
  }

  public Writer<?, ?> writeUint32(int value, Output<?> output) {
    return Base16.lowercase().writeIntLiteral(value, output, 8);
  }

  public int sizeOfUint64(long value) {
    return 18;
  }

  public Writer<?, ?> writeUint64(long value, Output<?> output) {
    return Base16.lowercase().writeLongLiteral(value, output, 16);
  }

  public int sizeOfBool(boolean value) {
    return value ? 4 : 5;
  }

  public Writer<?, ?> writeBool(boolean value, Output<?> output) {
    return Unicode.writeString(value ? "true" : "false", output);
  }

  public int sizeOfLambdaFunc(V bindings, V template) {
    return LambdaFuncWriter.sizeOf(this, bindings, template);
  }

  public Writer<?, ?> writeLambdaFunc(V bindings, V template, Output<?> output) {
    return LambdaFuncWriter.write(output, this, bindings, template);
  }

  public int sizeOfConditionalOperator(I ifTerm, I thenTerm, I elseTerm, int precedence) {
    return ConditionalOperatorWriter.sizeOf(this, ifTerm, thenTerm, elseTerm, precedence);
  }

  public Writer<?, ?> writeConditionalOperator(I ifTerm, I thenTerm, I elseTerm, int precedence, Output<?> output) {
    return ConditionalOperatorWriter.write(output, this, ifTerm, thenTerm, elseTerm, precedence);
  }

  public int sizeOfInfixOperator(I lhs, String operator, I rhs, int precedence) {
    return InfixOperatorWriter.sizeOf(this, lhs, operator, rhs, precedence);
  }

  public Writer<?, ?> writeInfixOperator(I lhs, String operator, I rhs, int precedence, Output<?> output) {
    return InfixOperatorWriter.write(output, this, lhs, operator, rhs, precedence);
  }

  public int sizeOfPrefixOperator(String operator, I operand, int precedence) {
    return PrefixOperatorWriter.sizeOf(this, operator, operand, precedence);
  }

  public Writer<?, ?> writePrefixOperator(String operator, I operand, int precedence, Output<?> output) {
    return PrefixOperatorWriter.write(output, this, operator, operand, precedence);
  }

  public int sizeOfInvokeOperator(V func, V args) {
    return InvokeOperatorWriter.sizeOf(this, func, args);
  }

  public Writer<?, ?> writeInvokeOperator(V func, V args, Output<?> output) {
    return InvokeOperatorWriter.write(output, this, func, args);
  }

  public abstract int sizeOfThen(V then);

  public abstract Writer<?, ?> writeThen(V then, Output<?> output);

  public int sizeOfIdentitySelector() {
    return 0;
  }

  public Writer<?, ?> writeIdentitySelector(Output<?> output) {
    return Writer.done();
  }

  public int sizeOfThenIdentitySelector() {
    return 0;
  }

  public Writer<?, ?> writeThenIdentitySelector(Output<?> output) {
    return Writer.done();
  }

  public int sizeOfLiteralSelector(I item, V then) {
    return LiteralSelectorWriter.sizeOf(this, item, then);
  }

  public Writer<?, ?> writeLiteralSelector(I item, V then, Output<?> output) {
    return LiteralSelectorWriter.write(output, this, item, then);
  }

  public int sizeOfThenLiteralSelector(I item, V then) {
    return 0;
  }

  public Writer<?, ?> writeThenLiteralSelector(I item, V then, Output<?> output) {
    return Writer.done();
  }

  public int sizeOfGetSelector(V key, V then) {
    return GetSelectorWriter.sizeOf(this, key, then);
  }

  public Writer<?, ?> writeGetSelector(V key, V then, Output<?> output) {
    return GetSelectorWriter.write(output, this, key, then);
  }

  public int sizeOfThenGetSelector(V key, V then) {
    return GetSelectorWriter.sizeOf(this, key, then);
  }

  public Writer<?, ?> writeThenGetSelector(V key, V then, Output<?> output) {
    return GetSelectorWriter.writeThen(output, this, key, then);
  }

  public int sizeOfGetAttrSelector(V key, V then) {
    return GetAttrSelectorWriter.sizeOf(this, key, then);
  }

  public Writer<?, ?> writeGetAttrSelector(V key, V then, Output<?> output) {
    return GetAttrSelectorWriter.write(output, this, key, then);
  }

  public int sizeOfThenGetAttrSelector(V key, V then) {
    return GetAttrSelectorWriter.sizeOf(this, key, then);
  }

  public Writer<?, ?> writeThenGetAttrSelector(V key, V then, Output<?> output) {
    return GetAttrSelectorWriter.writeThen(output, this, key, then);
  }

  public int sizeOfGetItemSelector(V index, V then) {
    return GetItemSelectorWriter.sizeOf(this, index, then);
  }

  public Writer<?, ?> writeGetItemSelector(V index, V then, Output<?> output) {
    return GetItemSelectorWriter.write(output, this, index, then);
  }

  public int sizeOfThenGetItemSelector(V index, V then) {
    return GetItemSelectorWriter.sizeOfThen(this, index, then);
  }

  public Writer<?, ?> writeThenGetItemSelector(V index, V then, Output<?> output) {
    return GetItemSelectorWriter.writeThen(output, this, index, then);
  }

  public int sizeOfKeysSelector(V then) {
    return KeysSelectorWriter.sizeOf(this, then);
  }

  public Writer<?, ?> writeKeysSelector(V then, Output<?> output) {
    return KeysSelectorWriter.write(output, this, then);
  }

  public int sizeOfThenKeysSelector(V then) {
    return KeysSelectorWriter.sizeOf(this, then);
  }

  public Writer<?, ?> writeThenKeysSelector(V then, Output<?> output) {
    return KeysSelectorWriter.writeThen(output, this, then);
  }

  public int sizeOfValuesSelector(V then) {
    return ValuesSelectorWriter.sizeOf(this, then);
  }

  public Writer<?, ?> writeValuesSelector(V then, Output<?> output) {
    return ValuesSelectorWriter.write(output, this, then);
  }

  public int sizeOfThenValuesSelector(V then) {
    return ValuesSelectorWriter.sizeOf(this, then);
  }

  public Writer<?, ?> writeThenValuesSelector(V then, Output<?> output) {
    return ValuesSelectorWriter.writeThen(output, this, then);
  }

  public int sizeOfChildrenSelector(V then) {
    return ChildrenSelectorWriter.sizeOf(this, then);
  }

  public Writer<?, ?> writeChildrenSelector(V then, Output<?> output) {
    return ChildrenSelectorWriter.write(output, this, then);
  }

  public int sizeOfThenChildrenSelector(V then) {
    return ChildrenSelectorWriter.sizeOf(this, then);
  }

  public Writer<?, ?> writeThenChildrenSelector(V then, Output<?> output) {
    return ChildrenSelectorWriter.writeThen(output, this, then);
  }

  public int sizeOfDescendantsSelector(V then) {
    return DescendantsSelectorWriter.sizeOf(this, then);
  }

  public Writer<?, ?> writeDescendantsSelector(V then, Output<?> output) {
    return DescendantsSelectorWriter.write(output, this, then);
  }

  public int sizeOfThenDescendantsSelector(V then) {
    return DescendantsSelectorWriter.sizeOf(this, then);
  }

  public Writer<?, ?> writeThenDescendantsSelector(V then, Output<?> output) {
    return DescendantsSelectorWriter.writeThen(output, this, then);
  }

  public int sizeOfFilterSelector(V predicate, V then) {
    return FilterSelectorWriter.sizeOf(this, predicate, then);
  }

  public Writer<?, ?> writeFilterSelector(V predicate, V then, Output<?> output) {
    return FilterSelectorWriter.write(output, this, predicate, then);
  }

  public int sizeOfThenFilterSelector(V predicate, V then) {
    return FilterSelectorWriter.sizeOfThen(this, predicate, then);
  }

  public Writer<?, ?> writeThenFilterSelector(V predicate, V then, Output<?> output) {
    return FilterSelectorWriter.writeThen(output, this, predicate, then);
  }

  public int sizeOfExtant() {
    return 0;
  }

  public Writer<?, ?> writeExtant(Output<?> output) {
    return Writer.done();
  }

  public int sizeOfAbsent() {
    return 0;
  }

  public Writer<?, ?> writeAbsent(Output<?> output) {
    return Writer.done();
  }
}

File: swim-js/swim-runtime/swim-core/swim-recon/src/main/writer/ReconWriter.ts
// Copyright 2015-2023 Swim.inc
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {Strings, Cursor} from "@swim/util";
import {Output, Writer, Unicode, Base10, Base16} from "@swim/codec";
import {Recon} from "../Recon";
import {AttrWriter} from "./AttrWriter";
import {SlotWriter} from "./SlotWriter";
import {BlockWriter} from "./BlockWriter";
import {PrimaryWriter} from "./PrimaryWriter";
import {MarkupTextWriter} from "./MarkupTextWriter";
import {DataWriter} from "./DataWriter";
import {IdentWriter} from "./IdentWriter";
import {StringWriter} from "./StringWriter";
import {LambdaFuncWriter} from "./LambdaFuncWriter";
import {ConditionalOperatorWriter} from "./ConditionalOperatorWriter";
import {InfixOperatorWriter} from "./InfixOperatorWriter";
import {PrefixOperatorWriter} from "./PrefixOperatorWriter";
import {InvokeOperatorWriter} from "./InvokeOperatorWriter";
import {LiteralSelectorWriter} from "./LiteralSelectorWriter";
import {GetSelectorWriter} from "./GetSelectorWriter";
import {GetAttrSelectorWriter} from "./GetAttrSelectorWriter";
import {GetItemSelectorWriter} from "./GetItemSelectorWriter";
import {KeysSelectorWriter} from "./KeysSelectorWriter";
import {ValuesSelectorWriter} from "./ValuesSelectorWriter";
import {ChildrenSelectorWriter} from "./ChildrenSelectorWriter";
import {DescendantsSelectorWriter} from "./DescendantsSelectorWriter";
import {FilterSelectorWriter} from "./FilterSelectorWriter";

/**
 * Factory for constructing Recon writers.
 * @public
 */
export abstract class ReconWriter<I, V> {
  abstract isField(item: I): boolean;

  abstract isAttr(item: I): boolean;

  abstract isSlot(item: I): boolean;

  abstract isValue(item: I): boolean;

  abstract isRecord(item: I): boolean;

  abstract isText(item: I): boolean;

  abstract isNum(item: I): boolean;

  abstract isBool(item: I): boolean;

  abstract isExpression(item: I): boolean;

  abstract isExtant(item: I): boolean;

  abstract items(item: I): Cursor<I>;

  abstract item(value: V): I;

  abstract key(item: I): V;

  abstract value(item: I): V;

  abstract string(item: I): string;

  abstract precedence(item: I): number;

  abstract sizeOfItem(item: I): number;

  abstract writeItem(output: Output, item: I): Writer;

  abstract sizeOfValue(value: V): number;

  abstract writeValue(output: Output, value: V): Writer;

  abstract sizeOfBlockValue(value: V): number;

  abstract writeBlockValue(output: Output, value: V): Writer;

  sizeOfAttr(key: V, value: V): number {
    return AttrWriter.sizeOf(this, key, value);
  }

  writeAttr(output: Output, key: V, value: V): Writer {
    return AttrWriter.write(output, this, key, value);
  }

  sizeOfSlot(key: V, value: V): number {
    return SlotWriter.sizeOf(this, key, value);
  }

  writeSlot(output: Output, key: V, value: V): Writer {
    return SlotWriter.write(output, this, key, value);
  }

  abstract sizeOfBlockItem(item: I): number;

  abstract writeBlockItem(output: Output, item: I): Writer;

  sizeOfBlock(item: I): number;
  sizeOfBlock(items: Cursor<I>, inBlock: boolean, inMarkup: boolean): number;
  sizeOfBlock(item: I | Cursor<I>, inBlock?: boolean, inMarkup?: boolean): number {
    if (arguments.length === 3) {
      return BlockWriter.sizeOf(this, item as Cursor<I>, inBlock!, inMarkup!);
    } else {
      const items = this.items(item as I);
      if (items.hasNext()) {
        return BlockWriter.sizeOf(this, items, this.isBlockSafe(this.items(item as I)), false);
      } else {
        return 2; // "{}"
      }
    }
  }

  writeBlock(output: Output, item: I): Writer;
  writeBlock(output: Output, items: Cursor<I>, inBlock: boolean, inMarkup: boolean): Writer;
  writeBlock(output: Output, item: I | Cursor<I>, inBlock?: boolean, inMarkup?: boolean): Writer {
    if (arguments.length === 4) {
      return BlockWriter.write(output, this, item as Cursor<I>, inBlock!, inMarkup!);
    } else {
      const items = this.items(item as I);
      if (items.hasNext()) {
        return BlockWriter.write(output, this, items, this.isBlockSafe(this.items(item as I)), false);
      } else {
        return Unicode.writeString(output, "{}");
      }
    }
  }

  sizeOfRecord(item: I): number {
    const items = this.items(item);
    if (items.hasNext()) {
      return BlockWriter.sizeOf(this, items, false, false);
    } else {
      return 2; // "{}"
    }
  }

  writeRecord(output: Output, item: I): Writer {
    const items = this.items(item);
    if (items.hasNext()) {
      return BlockWriter.write(output, this, items, false, false);
    } else {
      return Unicode.writeString(output, "{}");
    }
  }

  sizeOfPrimary(value: V): number {
    if (this.isRecord(this.item(value))) {
      const items = this.items(this.item(value));
      if (items.hasNext()) {
        return PrimaryWriter.sizeOf(this, items);
      }
    } else if (!this.isExtant(this.item(value))) {
      return this.sizeOfValue(value);
    }
    return 2; // "()"
  }

  writePrimary(output: Output, value: V): Writer {
    if (this.isRecord(this.item(value))) {
      const items = this.items(this.item(value));
      if (items.hasNext()) {
        return PrimaryWriter.write(output, this, items);
      }
    } else if (!this.isExtant(this.item(value))) {
      return this.writeValue(output, value);
    }
    return Unicode.writeString(output, "()");
  }

  isBlockSafe(items: Cursor<I>): boolean {
    while (items.hasNext()) {
      if (this.isAttr(items.next().value!)) {
        return false;
      }
    }
    return true;
  }

  isMarkupSafe(items: Cursor<I>): boolean {
    if (!items.hasNext() || !this.isAttr(items.next().value!)) {
      return false;
    }
    while (items.hasNext()) {
      if (this.isAttr(items.next().value!)) {
        return false;
      }
    }
    return true;
  }

  sizeOfMarkupText(item: I | string): number {
    if (typeof item !== "string") {
      item = this.string(item);
    }
    return MarkupTextWriter.sizeOf(item);
  }

  writeMarkupText(output: Output, item: I | string): Writer {
    if (typeof item !== "string") {
      item = this.string(item);
    }
    return MarkupTextWriter.write(output, item);
  }

  sizeOfData(length: number): number {
    return DataWriter.sizeOf(length);
  }

  writeData(output: Output, value: Uint8Array | undefined): Writer {
    if (value !== void 0) {
      return DataWriter.write(output, value);
    } else {
      return Unicode.writeString(output, "%");
    }
  }

  isIdent(value: I | string): boolean {
    if (typeof value !== "string") {
      value = this.string(value);
    }
    const n = value.length;
    let c: number | undefined;
    if (n === 0 || (c = value.codePointAt(0), c !== void 0 && !Recon.isIdentStartChar(c))) {
      return false;
    }
    for (let i = Strings.offsetByCodePoints(value, 0, 1); i < n; i = Strings.offsetByCodePoints(value, i, 1)) {
      c = value.codePointAt(i);
      if (c === void 0 || !Recon.isIdentChar(c)) {
        return false;
      }
    }
    return true;
  }

  sizeOfText(value: string): number {
    if (this.isIdent(value)) {
      return IdentWriter.sizeOf(value);
    } else {
      return StringWriter.sizeOf(value);
    }
  }

  writeText(output: Output, value: string): Writer {
    if (this.isIdent(value)) {
      return IdentWriter.write(output, value);
    } else {
      return StringWriter.write(output, value);
    }
  }

  sizeOfNum(value: number): number {
    if (isFinite(value) && Math.floor(value) === value && Math.abs(value) < 2147483648) {
      let size = Base10.countDigits(value);
      if (value < 0) {
        size += 1;
      }
      return size;
    } else {
      return ("" + value).length;
    }
  }

  writeNum(output: Output, value: number): Writer {
    if (isFinite(value) && Math.floor(value) === value && Math.abs(value) < 2147483648) {
      return Base10.writeInteger(output, value);
    } else {
      return Unicode.writeString(output, "" + value);
    }
  }

  sizeOfUint32(value: number): number {
    return 10;
  }

  writeUint32(output: Output, value: number): Writer {
    return Base16.lowercase.writeIntegerLiteral(output, value, 8);
  }

  sizeOfUint64(value: number): number {
    return 18;
  }

  writeUint64(output: Output, value: number): Writer {
    return Base16.lowercase.writeIntegerLiteral(output, value, 16);
  }

  sizeOfBool(value: boolean): number {
    return value ? 4 : 5;
  }

  writeBool(output: Output, value: boolean): Writer {
    return Unicode.writeString(output, value ? "true" : "false");
  }

  sizeOfLambdaFunc(bindings: V, template: V): number {
    return LambdaFuncWriter.sizeOf(this, bindings, template);
  }

  writeLambdaFunc(output: Output, bindings: V, template: V): Writer {
    return LambdaFuncWriter.write(output, this, bindings, template);
  }

  sizeOfConditionalOperator(ifTerm: I, thenTerm: I, elseTerm: I, precedence: number): number {
    return ConditionalOperatorWriter.sizeOf(this, ifTerm, thenTerm, elseTerm, precedence);
  }

  writeConditionalOperator(output: Output, ifTerm: I, thenTerm: I, elseTerm: I, precedence: number): Writer {
    return ConditionalOperatorWriter.write(output, this, ifTerm, thenTerm, elseTerm, precedence);
  }

  sizeOfInfixOperator(lhs: I, operator: string, rhs: I, precedence: number): number {
    return InfixOperatorWriter.sizeOf(this, lhs, operator, rhs, precedence);
  }

  writeInfixOperator(output: Output, lhs: I, operator: string, rhs: I, precedence: number): Writer {
    return InfixOperatorWriter.write(output, this, lhs, operator, rhs, precedence);
  }

  sizeOfPrefixOperator(operator: string, rhs: I, precedence: number): number {
    return PrefixOperatorWriter.sizeOf(this, operator, rhs, precedence);
  }

  writePrefixOperator(output: Output, operator: string, rhs: I, precedence: number): Writer {
    return PrefixOperatorWriter.write(output, this, operator, rhs, precedence);
  }

  sizeOfInvokeOperator(func: V, args: V): number {
    return InvokeOperatorWriter.sizeOf(this, func, args);
  }

  writeInvokeOperator(output: Output, func: V, args: V): Writer {
    return InvokeOperatorWriter.write(output, this, func, args);
  }

  abstract sizeOfThen(then: V): number;

  abstract writeThen(output: Output, then: V): Writer;

  sizeOfIdentitySelector(): number {
    return 0;
  }

  writeIdentitySelector(output: Output): Writer {
    return Writer.end();
  }

  sizeOfThenIdentitySelector(): number {
    return 0;
  }

  writeThenIdentitySelector(output: Output): Writer {
    return Writer.end();
  }

  sizeOfLiteralSelector(item: I, then: V): number {
    return LiteralSelectorWriter.sizeOf(this, item, then);
  }

  writeLiteralSelector(output: Output, item: I, then: V): Writer {
    return LiteralSelectorWriter.write(output, this, item, then);
  }

  sizeOfThenLiteralSelector(item: I, then: V): number {
    return 0;
  }

  writeThenLiteralSelector(output: Output, item: I, then: V): Writer {
    return Writer.end();
  }

  sizeOfGetSelector(key: V, then: V): number {
    return GetSelectorWriter.sizeOf(this, key, then);
  }

  writeGetSelector(output: Output, key: V, then: V): Writer {
    return GetSelectorWriter.write(output, this, key, then);
  }

  sizeOfThenGetSelector(key: V, then: V): number {
    return GetSelectorWriter.sizeOf(this, key, then);
  }

  writeThenGetSelector(output: Output, key: V, then: V): Writer {
    return GetSelectorWriter.writeThen(output, this, key, then);
  }

  sizeOfGetAttrSelector(key: V, then: V): number {
    return GetAttrSelectorWriter.sizeOf(this, key, then);
  }

  writeGetAttrSelector(output: Output, key: V, then: V): Writer {
    return GetAttrSelectorWriter.write(output, this, key, then);
  }

  sizeOfThenGetAttrSelector(key: V, then: V): number {
    return GetAttrSelectorWriter.sizeOf(this, key, then);
  }

  writeThenGetAttrSelector(output: Output, key: V, then: V): Writer {
    return GetAttrSelectorWriter.writeThen(output, this, key, then);
  }

  sizeOfGetItemSelector(index: V, then: V): number {
    return GetItemSelectorWriter.sizeOf(this, index, then);
  }

  writeGetItemSelector(output: Output, index: V, then: V): Writer {
    return GetItemSelectorWriter.write(output, this, index, then);
  }

  sizeOfThenGetItemSelector(index: V, then: V): number {
    return GetItemSelectorWriter.sizeOfThen(this, index, then);
  }

  writeThenGetItemSelector(output: Output, index: V, then: V): Writer {
    return GetItemSelectorWriter.writeThen(output, this, index, then);
  }

  sizeOfKeysSelector(then: V): number {
    return KeysSelectorWriter.sizeOf(this, then);
  }

  writeKeysSelector(output: Output, then: V): Writer {
    return KeysSelectorWriter.write(output, this, then);
  }

  sizeOfThenKeysSelector(then: V): number {
    return KeysSelectorWriter.sizeOf(this, then);
  }

  writeThenKeysSelector(output: Output, then: V): Writer {
    return KeysSelectorWriter.writeThen(output, this, then);
  }

  sizeOfValuesSelector(then: V): number {
    return ValuesSelectorWriter.sizeOf(this, then);
  }

  writeValuesSelector(output: Output, then: V): Writer {
    return ValuesSelectorWriter.write(output, this, then);
  }

  sizeOfThenValuesSelector(then: V): number {
    return ValuesSelectorWriter.sizeOf(this, then);
  }

  writeThenValuesSelector(output: Output, then: V): Writer {
    return ValuesSelectorWriter.writeThen(output, this, then);
  }

  sizeOfChildrenSelector(then: V): number {
    return ChildrenSelectorWriter.sizeOf(this, then);
  }

  writeChildrenSelector(output: Output, then: V): Writer {
    return ChildrenSelectorWriter.write(output, this, then);
  }

  sizeOfThenChildrenSelector(then: V): number {
    return ChildrenSelectorWriter.sizeOf(this, then);
  }

  writeThenChildrenSelector(output: Output, then: V): Writer {
    return ChildrenSelectorWriter.writeThen(output, this, then);
  }

  sizeOfDescendantsSelector(then: V): number {
    return DescendantsSelectorWriter.sizeOf(this, then);
  }

  writeDescendantsSelector(output: Output, then: V): Writer {
    return DescendantsSelectorWriter.write(output, this, then);
  }

  sizeOfThenDescendantsSelector(then: V): number {
    return DescendantsSelectorWriter.sizeOf(this, then);
  }

  writeThenDescendantsSelector(output: Output, then: V): Writer {
    return DescendantsSelectorWriter.writeThen(output, this, then);
  }

  sizeOfFilterSelector(predicate: V, then: V): number {
    return FilterSelectorWriter.sizeOf(this, predicate, then);
  }

  writeFilterSelector(output: Output, predicate: V, then: V): Writer {
    return FilterSelectorWriter.write(output, this, predicate, then);
  }

  sizeOfThenFilterSelector(predicate: V, then: V): number {
    return FilterSelectorWriter.sizeOfThen(this, predicate, then);
  }

  writeThenFilterSelector(output: Output, predicate: V, then: V): Writer {
    return FilterSelectorWriter.writeThen(output, this, predicate, then);
  }

  sizeOfExtant(): number {
    return 0;
  }

  writeExtant(output: Output): Writer {
    return Writer.end();
  }

  sizeOfAbsent(): number {
    return 0;
  }

  writeAbsent(output: Output): Writer {
    return Writer.end();
  }
}
